File: 2016/algorithm/Reverse String/reverseString-20160505.cc
#include <stack>
#include <iostream>
#include <string>

using namespace std;


// Solution 1
class Solution {
public: 
    string reverseString(string s) {
        stack<char> stk;
        for(unsigned i = 0; i < s.size(); i++) {
            stk.push(s[i]);
        }
        
        string ret = "";
        while(!stk.empty()) {
            ret += stk.top();
            stk.pop();
        }

        return ret;
    }
};


// Solution 2
class Solution2 {
public: 
    string reverseString(string s) {
       string ret = "";
       for(int i = s.size() - 1; i >= 0; i--) {
           ret += s[i];
       }

       return ret;
    }
};




int main() {
    Solution sol;
    Solution2 sol2;
    string concern = "angela";
    cout << sol.reverseString(concern) << "\n";
    cout << sol2.reverseString(concern) << "\n";
    return 0;
}

File: 2016/algorithm/344_Reverse_String/reverseString-20160505.py
# version 1, time limit exceeded
class Solution(object):
    def reverseString(self, s):
        ret = ""
        for i in range(len(s)):
            idx = len(s) - i - 1
            ret += s[idx]
        return ret


# version 2
class Solution2(object):
    def reverseString(self, s):
        return ''.join(reversed(s))


# version 3
class Solution3(object):
    def reverseString(self, s):
        return s[::-1]



concern = 'jesus'

sol = Solution()
print sol.reverseString(concern)


sol2 = Solution2()
print sol2.reverseString(concern)


sol3 = Solution3()
print sol3.reverseString(concern)
